New Logs Panel: Render new panel using the current visualization

matyax opened the pull request:
This PR updates the current Logs visualization to, when `newLogsPanel` is enabled, render the new and virtualized logs visualization. This allows to use the new panel by just enabling the feature flag, without having to make dashboard or application changes.

Changes and fixes:
- Integrated the new component to LogsPanel.
- Fixed rendering the icons for logs with errors, sampled logs, and deduplication count in unwrapped mode
- Increased overscan count (from 1 to 5). Decreases the empty space in the list when quickly scrolling, read more here: https://react-window.vercel.app/#/api/FixedSizeList
- When infinite scrolling is enabled, deduplication in `dataFrameToLogsModel()` is enabled, taking care of duplicated logs between the current response and the new response. This has been missing since we integrated it with the panel, and we do the same in Explore. https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/query.ts#L782
- Added support for custom menu options from apps, including dividers:
  <img width="326" alt="imagen" src="https://github.com/user-attachments/assets/183f2682-faed-4e19-870c-76f9b3e0b13f" />
- Enabled controls in Logs Sample Panel
- Made resize handler for Log Details more obvious


Part of #99075

matyax committed:
LogsPanel: integrate new panel via feature flag

matyax committed:
Log Line: move sampled/errors/deduplication count outside of log line body

matyax committed:
LogList: increase overscan count

matyax committed:
Logs Panel: enable deduplication for infinite scrolling

matyax committed:
Logs Panel: remove margin overflowing drilldown

matyax committed:
Logs Panel: add missing dependency to effect

matyax committed:
Logs Panel: pass missing callback

L2D2Grafana commented on public/app/features/logs/components/panel/virtualization.ts:
```diff
@@ -131,6 +131,7 @@ export function measureTextHeight(text: string, maxWidth: number, beforeWidth =
         delta += 1;
       } while (width >= availableWidth);
       if (beforeWidth) {
+        //console.log(testLogLine)
```
shush!

L2D2Grafana commented on public/app/features/logs/components/panel/LogLine.tsx:
```diff
@@ -334,20 +332,20 @@ export const getStyles = (theme: GrafanaTheme2) => {
     }),
     duplicates: css({
-      display: 'inline-block',
+      flexShrink: 0,
```
I wonder if this css is affecting the line overlap.

L2D2Grafana reviewed (commented):
This is niceee. I tried all the controls with the deduplication and trying to trigger infinite scroll at the end. Everything seemed as expected, except the line styling can overlap toggle wrap/unwrapp with deduplication on.

https://github.com/user-attachments/assets/cd3c61b6-e449-4b69-9955-2851021fe9ae

matyax committed:
Remove console log

matyax commented:
Thanks for testing @L2D2Grafana ! I'll see if I can reproduce that, no luck so far.

L2D2Grafana reviewed (approved):
It seems like it might just be me with some data from play, seeing the jitters.

gtk-grafana reviewed (commented):
Overall looking good, but a few bugs we should probably take a look at before merge:

Comments:
* Would like a smaller font option on service selection
<img width="1714" alt="image" src="https://github.com/user-attachments/assets/8c31d694-cccc-4707-872a-cf0cd9c7e41f" />

* "Inspect" that pulls up the line (or even the entire panel) in a modal might be nice in contexts like this so I can see a full log line outside of the container the panel is rendered in
* Would be nice to show the level color in logs details somewhere
* Missing expand/collapse JSON lines, enable/disable highlighting uses the same icon as expand/collapse JSON was using
* When selecting "show this field instead of the message" on a dervied field, would be nice to have the link in the log line 
* I don't like how the adaptive logs icon unaligns the start of the text 
<img width="182" alt="image" src="https://github.com/user-attachments/assets/dd57cff2-4ccc-4620-9f15-3e3cc8483493" />
<img width="229" alt="image" src="https://github.com/user-attachments/assets/b760f404-f9ce-4a78-af50-31c6001a2737" />

<hr />
Nits:

* Log details close button should not be able to be scrolled out of the details viewport
<img width="1706" alt="image" src="https://github.com/user-attachments/assets/0e775c32-3735-4daf-8587-9d26b3312422" />
* Logs are not vertically centered, there's more space below then above in the same line
<img width="1525" alt="image" src="https://github.com/user-attachments/assets/09323495-1bd0-4e61-b80b-d6a083ed4e77" />

SUPER NITS:
<hr />

Text overlaps when resizing, but like, this is such a silly trivial thing compared to how ****ing fast this panel renders. 
https://github.com/user-attachments/assets/6d29bb3c-dfbd-427f-af6f-9e3efb232990

<hr />
Bugs:

* Can no longer Show context or copy link to log line in Logs Drilldown
<img width="821" alt="image" src="https://github.com/user-attachments/assets/6aecd56f-2eee-497f-afb2-1666aefa1634" />
* Filters missing in log details in Logs Drilldown

PR: <img width="822" alt="image" src="https://github.com/user-attachments/assets/a888f02d-5414-4b8b-9655-23544c938982" />
Main: 
<img width="865" alt="image" src="https://github.com/user-attachments/assets/db726e50-2b90-4d96-aa49-e27f65ff0d66" />

* Toggling line wrap seems to trigger scroll, but not to the logs that were in the viewport before toggling the button.

https://github.com/user-attachments/assets/8f56c16f-9fac-41b5-8b36-96b8e8064929

* In explore, log details stays open even if the log line isn't in the results
<img width="1656" alt="image" src="https://github.com/user-attachments/assets/671000b0-24f1-4a50-bba9-a69c6d8bc0b6" />

* In dashboard panels "prettify JSON" toggles syntax highlighting

https://github.com/user-attachments/assets/ee33f15e-afdb-4fdc-a7fb-70163f4d61fa

matyax committed:
LogLine: show cursor pointer only when interactable

matyax committed:
LogLineDetails: make resize handler more obvious

matyax committed:
LogsPanel: add missing props to from panel

matyax committed:
LogLineMenu: add support for custom items

matyax committed:
LogsPanel: pass custom menu items to LogList

matyax committed:
Fix imports

matyax committed:
Chore: comments and missing argument

matyax committed:
LogLineMenu: pass log to event listener

matyax committed:
LogListContext: filter log details when no longer present in the new response

matyax committed:
chore: log

matyax committed:
LogsPanel: conditionally show options per feature flag status

matyax commented:
@gtk-grafana Amazing feedback as usual, thank you! I added the items to the epic for most comments and nits. A few answers below:

>    Missing expand/collapse JSON lines, enable/disable highlighting uses the same icon as expand/collapse JSON was using

I don't think this is coming back in the new panel for several reasons. One, the challenge of virtualizing, but more importantly it doesn't really make json more readable (at least for big payloads). The plan was to leverage details for that. It will be missed by some, but hopefully we can do something better in this iteration, that really addresses the challenge.

> Bugs:
> 
>     Can no longer Show context or copy link to log line in Logs Drilldown

Fixed. Missing required props.

Also open https://github.com/grafana/logs-drilldown/pull/1306 in Logs Drilldown.

> image * Filters missing in log details in Logs Drilldown

Fixed. Missing required props.

>     Toggling line wrap seems to trigger scroll, but not to the logs that were in the viewport before toggling the button.

Ahhh, I see. Not really scroll, but a reset of the virtualized panel. Not something that can be easily fixed in this PR, but I'll add it to the epic, as we also need to do some tweaking of the scroll position in some situations.

>     In explore, log details stays open even if the log line isn't in the results

I was unsure whether to keep it or not, but I'll take this feedback. We now filter open details when no longer present.


>     In dashboard panels "prettify JSON" toggles syntax highlighting

Oh yes, good call. Updated.

That should be all for now!

matyax committed:
LogLine: align logs when some of them are sampled or with errors

matyax committed:
Chore: update tests

gtk-grafana reviewed (approved):
Looks much better, but I'm running on <4 hours of sleep so probably missing something.

Oh, log level filters show up even if there aren't any logs

matyax committed:
LogLineMenu: test custom options

matyax committed:
LogsSamplePanel: show controls

matyax committed:
LogsPanel: move return after hooks to prevent bugs

matyax committed:
LogsSamplePanel: pass storage key
